### Pandas Lab - Finding, Querying, Creating Data

### Section 1: Selecting Data

Complete the following prompts, and compare your results with the answers to confirm that you did the operation properly.

In [2]:
pwd

'/Users/wzhang/GeneralAssembly/DAT-06-24/class material/Unit 2/Lesson 5/labs'

In [54]:
import pandas as pd
import numpy as np
df = pd.read_csv('../../data/titanic.csv')

In [4]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

**1).** Find the average age of all passengers on board.  The answer will be 29.7

In [5]:
# Your Code Here
df['Age'].mean()

29.69911764705882

In [21]:
df.Age.mean()

29.69911764705882

**2).** Find the median value of the 'Fare' and 'SibSp' columns.  The answer will be 14.45 and 0.00, respectively.

In [6]:
# Your Code Here
df[['Fare','SibSp']].median()

Fare     14.4542
SibSp     0.0000
dtype: float64

**3).** Find the median value of the 'Fare' and 'SibSp' columns in the first 100 rows.  The answer will be 15.675 and 0.00.

In [8]:
# Your Code Here
df[['Fare','SibSp']][:100].median()

Fare     15.675
SibSp     0.000
dtype: float64

**4).** Using the .iloc command, grab the modal value of the last 4 columns in the dataset.  The result should be a 3x3 dataframe that has the values 1601, 8.05, B96 B98, and S in the first row.

In [12]:
df.iloc[:,-4:].mode()

Ticket  Fare        Cabin Embarked
0      1601  8.05      B96 B98        S
1    347082   NaN  C23 C25 C27      NaN
2  CA. 2343   NaN           G6      NaN

**5).** Using the .iloc command, grab the mean value of the first 250 rows of the first 3 columns in the dataset.  The answer should be:
    - Passengerid: 125.5
    - Survived: 0.344
    - Pclass: 2.416

In [13]:
# Your Code Here
df.iloc[:250,:3].mean()

PassengerId    125.500
Survived         0.344
Pclass           2.416
dtype: float64

### Section 2: Querying Data

**1).** How many females were on board the Titanic? Men? The answers should be 314, 577

In [17]:
# Your Code Here
df[df.Sex == 'male']['PassengerId'].count()

577

In [35]:
df[df.Sex == 'male'].shape[0]

577

In [53]:
len(df[df.Sex == 'male'])

577

In [36]:
df[df.Sex == 'female'].shape[0]

314

In [24]:
df[df.Sex == 'female']['PassengerId'].count()

314

In [26]:
df[df.Sex == 'female']['PassengerId']

1        2
2        3
3        4
8        9
9       10
10      11
11      12
14      15
15      16
18      19
19      20
22      23
24      25
25      26
28      29
31      32
32      33
38      39
39      40
40      41
41      42
43      44
44      45
47      48
49      50
52      53
53      54
56      57
58      59
61      62
      ... 
807    808
809    810
813    814
816    817
820    821
823    824
829    830
830    831
835    836
842    843
849    850
852    853
853    854
854    855
855    856
856    857
858    859
862    863
863    864
865    866
866    867
871    872
874    875
875    876
879    880
880    881
882    883
885    886
887    888
888    889
Name: PassengerId, Length: 314, dtype: int64

In [32]:
df.loc[df.Sex == 'female', ['PassengerId']].count()

PassengerId    314
dtype: int64

In [30]:
type(df.loc[df.Sex == 'female', ['PassengerId']])

pandas.core.frame.DataFrame

**2).** What was the survival rate for females on the Titanic? The answer should be 0.74

In [39]:
# Your Code Here
df[(df.Sex=='female')]['Survived'].mean()

0.7420382165605095

In [47]:
df[(df.Sex=='female') & df.Survived == 1].shape[0] / df[df.Sex=='female'].shape[0]

0.7420382165605095

In [46]:
df[df.Sex=='female'].shape[0]

314

**3).** What was the survival rate for Males? The answer should be 0.19

In [48]:
# Your Code Here
df[(df.Sex=='male')]['Survived'].mean()

0.18890814558058924

**4).** What was the survival rate for passengers in either Pclass 1 or Pclass 2?  The answer should be 0.55

In [49]:
# Your Code Here
df[(df.Pclass==1) |(df.Pclass==2) ]['Survived'].mean()

0.5575

**5).** What was the survival rate if you were female and had at least 1 sibling on board? The answer should be 0.686.

In [52]:
# Your Code Here
df[(df.Sex=='female')&(df.SibSp > 0)]['Survived'].mean()

0.6857142857142857

### Section 3: Creating New Data

**1).** Create a column called 'Is_Alone' that's either True or False, depending on whether or not the column 'Family_Size' is > 0 or not.

In [60]:
# Your Code Here
df['Is_Alone'] = np.where((df.SibSp > 0 )| (df.Parch > 0), True, False)
# df.loc[:,['SibSp','Parch','Is_Alone']]
df[['SibSp','Parch','Is_Alone']]

SibSp  Parch  Is_Alone
0        1      0      True
1        1      0      True
2        0      0     False
3        1      0      True
4        0      0     False
5        0      0     False
6        0      0     False
7        3      1      True
8        0      2      True
9        1      0      True
10       1      1      True
11       0      0     False
12       0      0     False
13       1      5      True
14       0      0     False
15       0      0     False
16       4      1      True
17       0      0     False
18       1      0      True
19       0      0     False
20       0      0     False
21       0      0     False
22       0      0     False
23       0      0     False
24       3      1      True
25       1      5      True
26       0      0     False
27       3      2      True
28       0      0     False
29       0      0     False
..     ...    ...       ...
861      1      0      True
862      0      0     False
863      8      2      True
864      0      0     False
865      0      0     False
866      1      0      True
867      0      0     False
868      0      0     False
869      1      1      True
870      0      0     False
871      1      1      True
872      0      0     False
873      0      0     False
874      1      0      True
875      0      0     False
876      0      0     False
877      0      0     False
878      0      0     False
879      0      1      True
880      0      1      True
881      0      0     False
882      0      0     False
883      0      0     False
884      0      0     False
885      0      5      True
886      0      0     False
887      0      0     False
888      1      2      True
889      0      0     False
890      0      0     False

[891 rows x 3 columns]

**2).** Create a column called 'Demographic' that breaks people up into the following categories:
 - Below 8 years old: 'Child'
 - 8 - 21 years old: 'Adolescent'
 - 22 - 55 years old: 'Adult'
 - 55+: 'Senior'
 
When you're finished, use the method .value_counts() to confirm each section has the following count values:
 - Child: 50
 - Adolescent: 154
 - Adult: 470
 - Senior: 40

In [71]:
# Your Code Here
section_3_conditions = [
    df.Age < 8,
    df.Age <= 21,
    df.Age <= 55,
    df.Age > 55
]

section_3_result = ['Child', 'Adolescent', 'Adult', 'Senior']

df['Demographic'] = np.select(section_3_conditions, section_3_result, 'Not Set')

In [72]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Is_Alone Demographic  
0      0         A/5 21171   7.2500   NaN        S      True       Adult  
1      0          PC 17599  71.2833   C85        C      True       Adult  
2      0  STON/O2. 3101282   7.9250   NaN        S     False       Adult  
3      0            113803  53.1000  C123        S      True       Adult  
4      0            373450   8.0500   NaN        S     False       Adult

In [73]:
print(df[df.Demographic == 'Child'].shape[0])
print(df[df.Demographic == 'Adolescent'].shape[0])
print(df[df.Demographic == 'Adult'].shape[0])
print(df[df.Demographic == 'Senior'].shape[0])

50
154
470
40


In [75]:
df['Demographic'].value_counts()

Adult         470
Not Set       177
Adolescent    154
Child          50
Senior         40
Name: Demographic, dtype: int64

In [85]:
df.loc[df.Demographic == 'Child',['Age','Demographic']]

Age Demographic
7    2.00       Child
10   4.00       Child
16   2.00       Child
43   3.00       Child
50   7.00       Child
58   5.00       Child
63   4.00       Child
78   0.83       Child
119  2.00       Child
164  1.00       Child
171  4.00       Child
172  1.00       Child
183  1.00       Child
184  4.00       Child
193  3.00       Child
205  2.00       Child
233  5.00       Child
261  3.00       Child
278  7.00       Child
297  2.00       Child
305  0.92       Child
340  2.00       Child
348  3.00       Child
374  3.00       Child
381  1.00       Child
386  1.00       Child
407  3.00       Child
445  4.00       Child
448  5.00       Child
469  0.75       Child
479  2.00       Child
530  2.00       Child
535  7.00       Child
618  4.00       Child
642  2.00       Child
644  0.75       Child
691  4.00       Child
720  6.00       Child
750  4.00       Child
751  6.00       Child
755  0.67       Child
777  5.00       Child
788  1.00       Child
803  0.42       Child
813  6.00       Child
824  2.00       Child
827  1.00       Child
831  0.83       Child
850  4.00       Child
869  4.00       Child

In [87]:
df.loc[df.Demographic == 'Child','Age'].sum()

149.17000000000002

In [90]:
# Your Code Here
df['Demographic2'] = pd.cut(df['Age'], bins=[0, 7, 21, 55, df['Age'].max()], labels=['Child', 'Adolescent', 'Adult', 'Senior'])

In [91]:
df['Demographic2'].value_counts()

Adult         470
Adolescent    154
Child          50
Senior         40
Name: Demographic2, dtype: int64

**3).** Create a column called 'Gender_Status' that returns the following values:

 - 'F-High' if passenger is female and passenger class is 1.
 - 'F-Low' if passenger is female and passenger class is 2 or 3
 - 'M-High' if passenger is male and passenger class is 1
 - 'M-Low' if passenger is male and passenger class is 2 or 3
 
When you are finished, use .value_counts() to confirm you have the following count values for each of the following:
 - 'M-Low': 455
 - 'F-Low': 220
 - 'M-High': 122
 - 'F-High': 94

In [ ]:
# Your Code Here

**4).** Using string methods, extract the *greeting* of the persons name on board.  Ie, if someone's name is 'Ms. Madame Bovary', create a column that contains the value 'Ms.' and nothing else.

**Hint:** Take a look at the split() method and see what it does if you're not sure where to go.

In [ ]:
# Your Code Here